src/linux.rs

```diff
@@ -0,0 +1,354 @@
+} // Data follows
+
+/// Prepare an error for cases that "should never happen".
+fn unlikely_err(msg: String) -> Error {
+    debug_assert!(false, "{msg}");
+    Error::new(ErrorKind::Other, msg)
+}
+
+const fn addr_len(remote: &IpAddr) -> u8 {
+    match remote {
+        IpAddr::V4(_) => 32,
+        IpAddr::V6(_) => 128,
+    }
+}
+
+fn addr_bytes(remote: &IpAddr) -> Vec<u8> {
+    match remote {
+        IpAddr::V4(ip) => ip.octets().to_vec(),
+        IpAddr::V6(ip) => ip.octets().to_vec(),
+    }
+}
+
+const fn prepare_nlmsg(nlmsg_type: c_ushort, nlmsg_len: u32, nlmsg_seq: u32) -> libc::nlmsghdr {
+    let mut nlm = unsafe { zeroed::<nlmsghdr>() };
+    nlm.nlmsg_len = nlmsg_len;
+    nlm.nlmsg_type = nlmsg_type;
+    nlm.nlmsg_flags = NLM_F_REQUEST_U16 | NLM_F_ACK_U16;
+    nlm.nlmsg_seq = nlmsg_seq;
+    nlm
+}
+
+fn if_index(remote: IpAddr, fd: BorrowedFd) -> Result<i32, Error> {
+    // Prepare RTM_GETROUTE message.
+    #[allow(clippy::cast_possible_truncation)]
+    // Structs lens are <= u8::MAX per `const_assert!`s above; `addr_bytes` is max. 16 for IPv6.
+    let nlmsg_len = (size_of::<nlmsghdr>()
+        + size_of::<rtmsg>()
+        + size_of::<rtattr>()
+        + addr_bytes(&remote).len()) as u32;
+    #[allow(clippy::cast_sign_loss)]
+    // OK, because they are the same length and we just care about a unique value here.
+    let nlmsg_seq = fd.as_raw_fd() as u32;
+    let hdr = prepare_nlmsg(RTM_GETROUTE, nlmsg_len, nlmsg_seq);
+
+    let mut rtm = unsafe { zeroed::<rtmsg>() };
+    rtm.rtm_family = match remote {
+        IpAddr::V4(_) => AF_INET_U8,
+        IpAddr::V6(_) => AF_INET6_U8,
+    };
+    rtm.rtm_dst_len = addr_len(&remote);
+    rtm.rtm_table = RT_TABLE_MAIN;
+    rtm.rtm_scope = RT_SCOPE_UNIVERSE;
+    rtm.rtm_type = RTN_UNICAST;
+
+    let mut attr = unsafe { zeroed::<rtattr>() };
+    #[allow(clippy::cast_possible_truncation)]
+    // Structs len is <= u8::MAX per `const_assert!` above; `addr_bytes` is max. 16 for IPv6.
+    let rta_len = (size_of::<rtattr>() + addr_bytes(&remote).len()) as u16;
+    attr.rta_len = rta_len;
+    attr.rta_type = RTA_DST;
+
+    let mut buf = vec![0u8; nlmsg_len as usize];
+    unsafe {
+        ptr::copy_nonoverlapping(
+            ptr::from_ref(&hdr).cast(),
+            buf.as_mut_ptr(),
+            size_of::<nlmsghdr>(),
+        );
+        ptr::copy_nonoverlapping(
+            ptr::from_ref(&rtm).cast(),
+            buf.as_mut_ptr().add(size_of::<nlmsghdr>()),
+            size_of::<rtmsg>(),
+        );
+        ptr::copy_nonoverlapping(
+            ptr::from_ref(&attr).cast(),
+            buf.as_mut_ptr()
+                .add(size_of::<nlmsghdr>() + size_of::<rtmsg>()),
+            size_of::<rtattr>(),
+        );
+        ptr::copy_nonoverlapping(
+            addr_bytes(&remote).as_ptr(),
+            buf.as_mut_ptr()
+                .add(size_of::<nlmsghdr>() + size_of::<rtmsg>() + size_of::<rtattr>()),
+            addr_bytes(&remote).len(),
+        );
```

Comment:
You definitely don't want to be doing this manually. Either define a structure that arranges everything in memory as expected or write serialization routines for these structs.

Reply:
I was thinking about this more, and I think if you implement `Read` (and `Write`, where helpful) for each of the structures you're serializing, that alone will simplify a lot of this (similarly if you implement a helper struct that implements `Read`/`Write` for the `fd`).